unixfs/src/walk.rs

```diff
@@ -10,43 +10,183 @@ use std::convert::TryFrom;
 use std::fmt;
 use std::path::{Path, PathBuf};
 
-/// Representation of the current item of Walker or the last observed item with medatada.
+/// Representation of the walk progress. The common `Item` can be used to continue the walk.
 #[derive(Debug)]
-pub enum MetadataEntry<'a> {
-    /// Current item is a non-root plain Directory or a HAMTShard directory.
+pub enum ContinuedWalk<'a> {
+    /// Currently looking at a bucket.
+    Bucket(&'a Cid, &'a Path),
+    /// Currently looking at a directory.
     Directory(&'a Cid, &'a Path, &'a Metadata),
-    /// Current item is possibly a root file with a path, metadata, and a total file size.
-    File(&'a Cid, &'a Path, &'a Metadata, u64),
-    /// Current item is the root directory (HAMTShard or plain Directory).
+    /// Currently looking at a file. The first tuple value contains the file bytes accessible
+    /// from the block, which can also be an empty slice.
+    File(FileSegment<'a>, &'a Cid, &'a Path, &'a Metadata, u64),
+    /// Currently looking at a root directory.
     RootDirectory(&'a Cid, &'a Path, &'a Metadata),
-    /// Current item is possibly a root symlink.
-    Symlink(&'a Cid, &'a Path, &'a Metadata),
+    /// Currently looking at a symlink. The first tuple value contains the symlink target path. It
+    /// might be convertible to UTF-8, but this is not specified in the spec.
+    Symlink(&'a [u8], &'a Cid, &'a Path, &'a Metadata),
 }
 
-impl MetadataEntry<'_> {
-    /// Returns the metadata for the latest entry. It exists for initial directory entries, files,
-    /// and symlinks but not for continued HamtShards.
-    pub fn metadata(&self) -> &'_ Metadata {
+impl ContinuedWalk<'_> {
+    #[cfg(test)]
+    fn path(&self) -> &Path {
         match self {
-            Self::Directory(_, _, m)
-            | Self::File(_, _, m, ..)
-            | Self::RootDirectory(_, _, m)
-            | Self::Symlink(_, _, m) => m,
+            Self::Bucket(_, p)
+            | Self::Directory(_, p, ..)
+            | Self::File(_, _, p, ..)
+            | Self::RootDirectory(_, p, ..)
+            | Self::Symlink(_, _, p, ..) => p,
+        }
+    }
+}
+
+/// Errors which can occur while walking a tree.
+#[derive(Debug)]
+pub enum Error {
```

Comment:
Moving `Error` and `ContinuedWalk`, `FileSegment` back to where it was would certainly drop the lines changed.